Add coding guidelines on C standard library function usage

stephanosio opened the pull request:
This series adds the coding guidelines rules on the usage of C standard library functions in the Zephyr codebase, in preparation for making Picolibc the default libc.

It adds two new rules:

* Rule A.4: C Standard Library Usage Restrictions in Zephyr Kernel
    * Ensures that the Zephyr kernel can still build with the minimal libc even after making Picolibc the default.
* Rule A.5: C Standard Library Usage Restrictions in Zephyr Codebase
    * Prohibit the use of non-standard (non-ISO) functions from Picolibc (or any other LIBCs for that matter) in the Zephyr codebase

Once these rules are accepted, additional testcases and CI checks will be implemented to enforce them.

Preview: https://builds.zephyrproject.io/zephyr/pr/57598/docs/index.html

Part of https://github.com/zephyrproject-rtos/zephyr/issues/49922

stephanosio committed:
doc: languages: c: Add minimal libc function scope description

This commit adds the "Functions" section to the minimal libc
documentation that describes the scope of the C standard library
functions available in the minimal libc.

Signed-off-by: Stephanos Ioannidis <stephanos.ioannidis@nordicsemi.no>

stephanosio committed:
doc: Add kernel C standard library usage restriction rule

This commit adds the "Rule A.4: C Standard Library Usage Restrictions
in Zephyr kernel" to the coding guidelines.

Signed-off-by: Stephanos Ioannidis <stephanos.ioannidis@nordicsemi.no>

stephanosio committed:
doc: Add Zephyr codebase C standard library usage restriction rule

This commit adds the "Rule A.5: C Standard Library Usage Restrictions
in Zephyr Codebase" to the coding guidelines.

The initial list of allowed non-ISO C libc functions is based on the
non-ISO C functions available in the minimal libc.

Signed-off-by: Stephanos Ioannidis <stephanos.ioannidis@nordicsemi.no>

galak commented on doc/contribute/coding_guidelines/index.rst:
```diff
@@ -1297,6 +1297,85 @@ macros. Given that Zephyr uses a fork of the corresponding upstream for each
 module, it is always possible to patch the macro implementation in each module
 to avoid collisions.
 
+.. _coding_guideline_libc_usage_restrictions_in_zephyr_kernel:
```
We should put something in here that any exception should require an implementation in `lib/libc/common` and test in `tests/lib/c_lib/` so as to provide an implementation that can be used for out of tree libc's that may not support the function.

stephanosio replied:
Added the following:

```
All non-ISO functions and macros listed above must be implemented by the
:ref:`common libc <c_library_common>` in order to make sure that these
functions can be made available when using a C standard library that does not
implement these functions.
```

keith-packard commented:
I'm working to make Picolibc expose the "Zephyr" C library API directly. This will also enable the removal of `_POSIX_C_SOURCE` definition when building with picolibc, which should ensure that core Zephyr doesn't use API from outside the accepted set.

keith-packard commented on doc/contribute/coding_guidelines/index.rst:
```diff
@@ -1297,6 +1297,101 @@ macros. Given that Zephyr uses a fork of the corresponding upstream for each
+-----------
+
+Zephyr kernel shall not use the C standard library functions and macros that
+are not available in the :ref:`minimal libc <c_library_minimal>`.
```
I'd phrase this as using the standard C library as defined in ISO/IEC 9899:2018 plus a set of extensions (named here, identifying which standard they are defined by). That way we have both the list of functions *and* explicit specifications for them.

keith-packard replied:
In particular, I don't think this set of rules should refer to the minimal C library, it should define the Zephyr C library directly.

stephanosio replied:
At least for the Rule A.4, its purpose is "to ensure that the Zephyr kernel can build with the minimal libc," so I would prefer to explicitly refer to the set of functions available in the minimal libc.

As for the Rule A.5, which basically defines the "Zephyr C library API," I agree with you and will update the list to define the "specifications" for them.

keith-packard replied:
> At least for the Rule A.4, its purpose is "to ensure that the Zephyr kernel can build with the minimal libc," so I would prefer to explicitly refer to the set of functions available in the minimal libc.

That makes sense -- explicitly define the set of functions currently allowed for use by the Zephyr kernel. However, I think it should be worded the other way around -- define the set of functions allowed within the Zephyr kernel and then specify that the minimal C library must support all of those functions. When a new function is proposed for use within the kernel, this form of the specification would naturally drive the requirement that the PR include an implementation within the minimal C library.

keith-packard commented on doc/contribute/coding_guidelines/index.rst:
```diff
@@ -1297,6 +1297,85 @@ macros. Given that Zephyr uses a fork of the corresponding upstream for each
+Zephyr codebase shall not use the functions and macros that are not defined by
+the ISO C standard, unless exempted by this rule.
+
+The "Zephyr codebase" in this context refers to all source code files committed
```
Yup, seems like an exception needs to be added which allows use of functions from the POSIX extensions to the C library as defined by IEEE Std 1003.1-2017.

cfriedt replied:
> Yup, seems like an exception needs to be added which allows use of functions from the POSIX extensions to the C library as defined by IEEE Std 1003.1-2017.

And presumably, those should be on an "approved" list as well.

stephanosio replied:
The scope of this is the functions and macros **from the libc** as implied by the name of this rule: "Rule A.5: C Standard Library Usage Restrictions in Zephyr Codebase."

So basically, "Zephyr codebase shall not use the **C standard library-provided** functions and macros that are not defined by the ISO C standard, unless exempted by this rule."

Since the Zephyr POSIX library functions are provided by the Zephyr POSIX library (i.e. not libc/C standard library), that is outside the scope of this rule and therefore not disallowed by this rule.

I will try to reword this to make that more clear.

keith-packard replied:
Perhaps we need a separate specification for which POSIX functions are supported by the Zephyr POSIX extension?

stephanosio replied:
> Perhaps we need a separate specification for which POSIX functions are supported by the Zephyr POSIX extension?

https://docs.zephyrproject.org/latest/services/portability/posix.html#units-of-functionality might be what you are looking for.

keith-packard replied:
yup, sorry for not looking around.

keith-packard commented on doc/contribute/coding_guidelines/index.rst:
```diff
@@ -1297,6 +1297,101 @@ macros. Given that Zephyr uses a fork of the corresponding upstream for each
+to the `main Zephyr repository`_, except the Zephyr kernel as defined by the
+:ref:`coding_guideline_libc_usage_restrictions_in_zephyr_kernel`.
+
+The following non-ISO functions and macros are exempt from this rule and may be
```
I think this is somewhat awkwardly specified. I think we should define the Zephyr C library API as being a combination of the ISO/IEC standard C library along with a few functions borrowed from other standards.

stephanosio replied:
https://github.com/zephyrproject-rtos/zephyr/pull/57598#discussion_r1190272409

I will update the Rule A.5 to explicitly refer to the ISO/IEC 9899.

stephanosio replied:
Updated the coding guidelines to refer to to the ISO/IEC 9899:2011.

keith-packard commented on doc/develop/languages/c/minimal_libc.rst:
```diff
@@ -33,6 +33,90 @@ The minimal libc uses the malloc api family implementation provided by the
+Functions
+*********
+
+The minimal libc implements only the minimal subset of the standard C library
```
I wonder if it would be reasonable to allow any ISO/IEC C library function to be used by Zephyr code so that developers weren't encouraged to re-implement them slightly incorrectly on their own? A patch adding use of a new function would need to include an implementation for use in the minimal C library. This seems aligned with the goal of continuing to support external ISO/IEC compliant toolchains while not hindering use of well defined standard functions.

stephanosio replied:
> I wonder if it would be reasonable to allow any ISO/IEC C library function to be used by Zephyr code so that developers weren't encouraged to re-implement them slightly incorrectly on their own?

That is the goal of Rule A.5, which says "Zephyr codebase shall not use the functions and macros that are not defined by the ISO C standard" (i.e. it does not say Zephyr codebase only needs to use the ISO C standard functions that are implemented by the Zephyr minimal libc -- that only applies to the "Zephyr kernel," which is a small subset of Zephyr).

I will try to reword the Rule A.5 to make this more clear.

stephanosio replied:
Done (hopefully).

stephanosio commented on doc/contribute/coding_guidelines/index.rst:
```diff
@@ -1297,6 +1297,101 @@ macros. Given that Zephyr uses a fork of the corresponding upstream for each
+* OS Library (:file:`lib/os`)
+* Architecture Port (:file:`arch`)
+* Logging Subsystem (:file:`subsys/logging`)
+
```
> drivers/ and other subsys/ are two areas that I think this should generally apply.

Now that is almost the entirety of Zephyr and doing so defeats the purpose of making Picolibc the default libc, which is to allow more C standard library functions to be used throughout the Zephyr codebase.

As per https://github.com/zephyrproject-rtos/zephyr/issues/49922#issuecomment-1251292117, the minimal libc was originally intended "to support kernel functionality and test cases" and "it is not designed to be used with applications."

The goal is to make sure that the "Zephyr kernel" (as in the bare minimum core parts of Zephyr required to test what is under `kernel/`) can keep building with the minimal libc for "self-contained testing and verification of the kernel and core OS services" -- that certainly should not require every subsystem under `subsys/` or every driver under `drivers/`.

keith-packard replied:
Any reason for sticking the non ISO-C functions in backticks? Or are they links to the relevant specs that I'm just not seeing here?

keith-packard replied:
Are you considering all of arch/ to be part of this list? That seems almost as unbounded as drivers/

keith-packard replied:
I'm wondering if this is a good place to note that patches adding use of additional ISO/IEC C functions are welcome as long as they are accompanied by implementations of those functions for the minimal C library. I'd like it to be clear that the goal is *not* to restrict the usage of standard functions unnecessarily, just to ensure that the repository can be built without use of external code.

keith-zephyr replied:
> Are you considering all of arch/ to be part of this list? That seems almost as unbounded as drivers/

What about restricting to arch/common and arch/posix?

keith-packard reviewed (changes requested):
Yup, this is looking good. I'd like to see a policy about extending the set of functions allowed here .

keith-zephyr commented on doc/contribute/coding_guidelines/index.rst:
```diff
@@ -1297,6 +1297,188 @@ macros. Given that Zephyr uses a fork of the corresponding upstream for each
+-----------
+
+The use of the C standard library functions and macros in the Zephyr kernel
+shall be limited to the following functions and macros from the ISO/IEC
+9899:2011 standard, also known as C11, and their extensions:
```
Regarding ISO/IEC 9899:2011 standard.  This standard has been [withdrawn](https://www.iso.org/standard/57853.html) and replaced by ISO/IEC 9899:2018.

Do we need a doc scrub as a separate PR to replace 9899:2011 references with 9899:2018?

keith-packard replied:
I think we're free to reference the older spec if we please; it's still well defined, even if not current. We should consider when a transition to :2018 is to be made.

stephanosio replied:
The reference to the ISO/IEC 9899:2011 (aka. C11) is intentional since the project is still adhering to the ISO/IEC 9899:1999 (aka. C99) for the most part and, at most, C11 (see https://github.com/zephyrproject-rtos/zephyr/issues/30105).

While C17 is technically only bug fixes for the C11, it is still a different version and some toolchains that the Zephyr project and its downstream users depend on may not have the support for it.

carlescufi commented:
Architecture WG:

- @fabiobaltieri asks how to enforce A.4: We will build the "zephyr kernel" against the minimal libc to ensure those don't use any C standard lib functions outside the list
- @nashif asks why the logging subsystem is included. @stephanosio replies that the kernel uses the logging subsystem
- @evgeniy-paltsev asks why would we allow something like `malloc` in the kernel. @stephanosio replies that this is only about the C standard library, not whether the function makes sense in the kernel
- @nashif states that he agrees with @evgeniy-paltsev that things like `malloc` or `qsort` don't make sense in the kernel.
- @stephanosio presents rule A.5, that lists the functions that are non-standard but allowed in the zephyr codebase

keith-packard commented on doc/contribute/coding_guidelines/index.rst:
```diff
@@ -1297,6 +1297,188 @@ macros. Given that Zephyr uses a fork of the corresponding upstream for each
+   putchar(),ISO/IEC 9899:2011
+   puts(),ISO/IEC 9899:2011
+   qsort(),ISO/IEC 9899:2011
+   `qsort_r()`_,GNU
```
That leaves `strtok_r` and `strnlen` as the only non ISO-C functions in this list. `strnlen` is used once in lib/os/cbprintf_complete.c (which is only used if `CONFIG_CBPRINTF_COMPLETE` is defined). There's a spurious prototype in that file because it doesn't define `_POSIX_C_SOURCE 200809`. `strtok_r` does not appear to be used at all. Looks like `strtok_r` still needs to get moved to lib/libc/common.

stephanosio replied:
> Sounds like I should remove them from the `_ZEPHYR_SOURCE_` set in picolibc as well.

For now, yes for the purpose of simplicity and consistency.

Note, however, that the set of non-standard functions allowed to be used in the Zephyr codebase (i.e. outside the "Zephyr kernel") does not necessarily have to match the set of non-standard functions allowed to be used in the "Zephyr kernel" given that the latter is a subset of the former.

> `strtok_r` does not appear to be used at all.

`strtok_r` is used in some drivers and subsystems but not by any of the components that belong in the "Zephyr kernel" scope, so it could technically be removed from the Rule A.4.

This is actually where "the latter is a subset of the former" case comes in -- the Rule A.5 (and subsequently Picolibc with `_ZEPHYR_SOURCE_`) includes `strtok_r`, but not the Rule A.4.

> Looks like strtok_r still needs to get moved to lib/libc/common.

Correct.

keith-packard replied:
Ok, so let me make sure I understand how this should work in practice.

Code *outside* of the "Zephyr kernel" may use functions defined in POSIX, BSD or GNU standards by defining the relevant `_*_SOURCE` macro at the top of the source file. Such code could only be used when building with a C library supporting the relevant standard. Code *inside* the "Zephyr kernel" _must not_ define any `_*_SOURCE` macro and the build environment will ensure that the visible API does not include anything from those standards which is not explicitly called out.

Is it necessary for the C library to expose only the "Zephyr kernel API" when being used in this "Zephyr mode"? Or is it acceptable for the C library to expose a complete ISO/IEC C API along with the Zephyr extensions derived from the other standards? I'm expecting the latter to be the answer as the former would require significant changes to each C library, right?

stephanosio replied:
> Code _outside_ of the "Zephyr kernel" may use functions defined in POSIX, BSD or GNU standards by defining the relevant `_*_SOURCE` macro at the top of the source file.

No, code outside of the "Zephyr kernel" is subject to the Rule A.5 and may only use the non-standard functions listed in that rule, which will be enforced by the means of `_ZEPHYR_SOURCE_`; no other feature test macro should be used anywhere in the upstream Zephyr codebase.

What I am saying is that the Rule A.5 (i.e. outside of the "Zephyr kernel") may contain some non-standard functions that are not part of the Rule A.4 (i.e. inside the "Zephyr kernel").

One example is `strtok_r`, which will _not_ be listed in the Rule A.4 and made `default n` for the minimal libc so that it cannot be used inside the "Zephyr kernel"; but, it will still be listed in the Rule A.5 and made part of `_ZEPHYR_SOURCE_` so that it can be used outside the "Zephyr kernel."

keith-packard replied:
> No, code outside of the "Zephyr kernel" is subject to the Rule A.5 and may only use the non-standard functions listed in that rule, which will be enforced by means of `_ZEPHYR_SOURCE_`; no other feature test macro should be used anywhere in the upstream Zephyr codebase (inside or outside the "Zephyr kernel").

Hrm. So code inside the Zephyr codebase may use any new ISO/IEC C library function by providing an implementation for the minimal C library, but it may only use a new C library function from another standard by first getting permission from the TSC to extend the set of allowed functions, providing an implementation for the common C library and then arranging for this new function to be made available under `_ZEPHYR_SOURCE_`, both for the minimal C library as well as every other C library the code might be compiled against. 

In the case of picolibc, that will involve providing patches to picolibc to adjust the definition of `_ZEPHYR_SOURCE_` and then updating both the SDK and the module, as well as providing backwards-compatibility hacks within Zephyr to support older SDK versions. Compared with having internal code directly specify the relevant standard API suite they wish to use (by setting the appropriate `_*_SOURCE` macro), this seems like a pretty significant burden. 

I fear this will discourage developers from using standard functions and push them towards re-implementing them instead. I believe it's more important to ensure use of well documented and tested implementations of standard functionality than to try and eliminate references to external standards in the source code -- we've already seen that we have to refer to those external standards in the documentation anyways to ensure that the definition of the Zephyr C library API is accurate.

In other words, I think we've made some incorrect design choices here. We've got the right goal in mind -- define the Zephyr C API accurately; control the set of C library functions used by various layers of Zephyr code closely; provide a process for adding new C library functions that encourages developers to use standard functions as much as practical. However, we're trying too hard to hide the actual standards from which these functions are derived with the result being a divergence from the underlying standards as well as layers of kludges within the Zephyr code base.

stephanosio replied:
> Hrm. So code inside the Zephyr codebase may use any new ISO/IEC C library function by providing an implementation for the minimal C library

No. Outside the "Zephyr kernel," you are allowed to use any ISO C standard function, as stated in the Rule A.5.

```
The use of the C standard library functions and macros in the Zephyr codebase
shall be limited to the functions from the ISO/IEC 9899:2011 standard, also
known as C11, unless exempted by this rule.
```

> but it may only use a new C library function from another standard by first getting permission from the TSC to extend the set of allowed functions, providing an implementation for the common C library and then arranging for this new function to be made available under _ZEPHYR_SOURCE_

That only applies to non-standard functions.

I think you are misunderstanding and mixing up the Rule A.4 and A.5. I would suggest re-reading these rules as proposed this PR carefully again.

stephanosio replied:
I will try to summarise this so it is less confusing:

**Inside "Zephyr kernel"** 

* subject to the Rule A.4.
* only the functions explicitly listed in the Rule A.4 may be used (applies to both standard and non-standard functions).
* all functions listed in the Rule A.4 must be supported by the minimal libc.
    * all standard functions listed in the Rule A.4 must be implemented by the minimal libc.
    * all non-standard functions listed in the Rule A.4 must be implemented by the common libc.

**For the rest of the Zephyr codebase (i.e. outside "Zephyr kernel")**

* subject to the Rule A.5.
* all standard functions from the ISO/IEC 9899:2011 standard may be used (since they will be implemented by all standard-conforming LIBCs).
* non-standard functions may not be used unless explicitly listed in the Rule A.5.
    * all non-standard functions listed in the Rule A.5 must implemented by the common libc.

keith-packard replied:
Yes, that's helpful, and I think I understand the rules for code *using* C library functions, but I'm a lot more interested in the C library implementation itself and how it is supposed to advertise these functions, along with how we extend the lists of functions in both Rules A.4 and A.5.

Adding functions to the list in A.4 seems pretty easy -- provide an implementation of the function for the minimal C library and then just start using it. You'd need to get approval during the PR process. Any other C library will already provide an implementation and API declaration under the existing compilation rules, so there's nothing needed to use the new function with them.

Adding functions under A.5 seems more complicated, and potentially hard enough to discourage people. Similar to A.4, you need to provide an implementation for the common C library and get approval during the PR process. However, to start using this with any other C library, you need to provide a declaration of the API for every supported C library using the `#include_next` mechanism that is being proposed in #57619, even for libraries which support the relevant standard (be that BSD, GNU or POSIX). For these libraries, there is a potential for skew between the Zephyr internal declarations and the library, which could lead to an accidental ABI difference and subsequent application failures, either at compile time or run time.

I'd like to consider a slightly modified version of #57619. 

 * Where the external library does not provide a definition of the relevant API, we have a Zephyr-specific version of the relevant header which `#include`s a common C library headers declaring the requested APIs along with a `#include_next` to pull in the external library version of the file (although, that's problematic given that `#include_next` is a GCC extension).
 * Where the external library *does* provide a definition of the relevant API, there is no Zephyr-specific version of the relevant header. Instead, the source code using the function would include the necessary `#define _*_SOURCE` statement so that the external library headers would include the declaration of the desired API. This would avoid any possibility of ABI skew between the Zephyr declaration and the library definition.
 * When using the picolibc module, that code would always be synchronized with the set of allowed APIs in A.5 under `_ZEPHYR_SOURCE`, so the `#define _*_SOURCE` statement would not be required.

Making the `#define _*_SOURCE` conditional upon whether the picolibc module was being used would allow verification of the source code to ensure it wasn't using functions beyond the scope of these rules without also adding a layer of non-standard C preprocessor usage -- the verification would only occur when using the picolibc module or C library that doesn't provide an API beyond ISO/IEC C.

Hrm. Maybe we should just create header files that declare the functions in A.5? Something like `include/zephyr/libc.h`, or even `include/zephyr/string.h` et al. We could hide all of the library-specific goo there without relying on `#include_next`? That could check for the relevant config entries pulling implementations out of the common C library and insert declarations for those functions. For picolibc, `_ZEPHYR_SOURCE_` could then *omit* declarations for functions that are pulled from the common C library (malloc, et al). This would catch use of malloc family functions not provided in the common C library implementation.

With the current plan, I don't see any reason to include support for `_ZEPHYR_SOURCE` in picolibc -- Zephyr will always provide wrapping include files that declare the Zephyr API on top of the picolibc one, code for this in picolibc would be redundant.

stephanosio replied:
> Adding functions under A.5 seems more complicated, and potentially hard enough to discourage people.

It can be a bit more involved than A.4 indeed; though, this only applies to non-standard functions since the standard functions are allowed by default by A.5.

> Similar to A.4, you need to provide an implementation for the common C library and get approval during the PR process.

Correct.

> However, to start using this with any other C library, you need to provide a declaration of the API for every supported C library using the `#include_next` mechanism that is being proposed in #57619, even for libraries which support the relevant standard (be that BSD, GNU or POSIX).

That would be the case for the libraries that cannot be modified to support a Zephyr-specific feature test macro (e.g. `_ZEPHYR_SOURCE`).

> For these libraries, there is a potential for skew between the Zephyr internal declarations and the library, which could lead to an accidental ABI difference and subsequent application failures, either at compile time or run time.

Correct. But, we require tests for all functions added to our LIBCs, so any "accidental ABI difference" that is a real problem would be caught by such tests.

> I'd like to consider a slightly modified version of https://github.com/zephyrproject-rtos/zephyr/pull/57619.

> * Where the external library does not provide a definition of the relevant API, we have a Zephyr-specific version of the relevant header which `#include`s a common C library headers declaring the requested APIs along with a `#include_next` to pull in the external library version of the file (although, that's problematic given that `#include_next` is a GCC extension).

Yes, that sounds reasonable.

Re: `#include_next`, it is one of the few extensions that we take for granted (we should codify and document these one day ...). The expectation is that all C compilers (preprocessors), even for its own sake, should support `#include_next` or have something equivalent to it since it would be pretty hard to write a C library without such a directive. Even IAR, the infamous compiler that is very conservative and lacks the support for various de facto standard C extensions, supports `#include_next`.

> * Where the external library _does_ provide a definition of the relevant API, there is no Zephyr-specific version of the relevant header. Instead, the source code using the function would include the necessary `#define _*_SOURCE` statement so that the external library headers would include the declaration of the desired API. This would avoid any possibility of ABI skew between the Zephyr declaration and the library definition.

IMHO, this requires a further discussion. I have no problem with either approach (declaring local prototypes vs. using the feature test macros) for the C libraries other than Picolibc since they are not used for enforcing the Rule A.5 in our CI.

Some points I want to add is:

1. Your concern about the ABI skew between the Zephyr declaration and the library declaration should be addressed by the C library test.
2. If we decide to go with using the feature test macros, the macros must _not_ be defined by each source file using the functions covered by such macros; instead, these macros should be defined by the libc integration headers under `lib/libc/*/include`. The rationale is that the feature test macros like `_GNU_SOURCE` are not standardised and may vary across different LIBCs, and we do not want to clutter the general source files in the Zephyr codebase with these LIBC-specific feature test macros.

> * When using the picolibc module, that code would always be synchronized with the set of allowed APIs in A.5 under `_ZEPHYR_SOURCE`, so the `#define _*_SOURCE` statement would not be required.

Correct.

> With the current plan, I don't see any reason to include support for `_ZEPHYR_SOURCE` in picolibc -- Zephyr will always provide wrapping include files that declare the Zephyr API on top of the picolibc one, code for this in picolibc would be redundant.

If you are referring to the changes in #57619, the `lib/libc/picolibc/include/*.h` providing the Zephyr-side prototype declarations for the non-standard functions is only a temporary workaround until we update the minimum required Zephyr SDK version to a version that contains a more up-to-date Picolibc with `_ZEPHYR_SOURCE` support -- once the minimum required Zephyr SDK version is updated, the Picolibc workarounds can be removed entirely and we can rely on `_ZEPHYR_SOURCE`.

> Hrm. Maybe we should just create header files that declare the functions in A.5? Something like `include/zephyr/libc.h`, or even `include/zephyr/string.h` et al. We could hide all of the library-specific goo there without relying on `#include_next`? That could check for the relevant config entries pulling implementations out of the common C library and insert declarations for those functions. For picolibc, `_ZEPHYR_SOURCE_` could then _omit_ declarations for functions that are pulled from the common C library (malloc, et al). This would catch use of malloc family functions not provided in the common C library implementation.

I am not sure if I understand what you are suggesting here.

Are you saying that, for instance, we define our own `string.h` with all standard and non-standard `string.h` functions so that we do not need to include the LIBC `string.h` at all?

But either way, as I explained above, we currently take `#include_next` for granted and the LIBC would not be the first thing to fail if we come across a compiler that does not support `#include_next`, so I would not worry too much about

keith-packard replied:
> > However, to start using this with any other C library, you need to provide a declaration of the API for every supported C library using the `#include_next` mechanism that is being proposed in #57619, even for libraries which support the relevant standard (be that BSD, GNU or POSIX).
> 
> That would be the case for the libraries that cannot be modified to support a Zephyr-specific feature test macro (e.g. `_ZEPHYR_SOURCE`).

Not entirely -- we will need to add support in Zephyr to support older SDK versions and/or external versions of said libraries. Just as #57619 is adding picolibc support.

> Re: `#include_next`, it is one of the few extensions that we take for granted (we should codify and document these one day ...). The expectation is that all C compilers (preprocessors), even for its own sake, should support `#include_next` or have something equivalent to it since it would be pretty hard to write a C library without such a directive. Even IAR, the infamous compiler that is very conservative and lacks the support for various de facto standard C extensions, supports `#include_next`.

Good to hear (btw, picolibc doesn't use `#include_next` internally, but glibc sure makes a lot of use of it).

>     1. Your concern about the ABI skew between the Zephyr declaration and the library declaration should be addressed by the C library test.

Sounds like there's a requirement to add tests within Zephyr for all functions added in this fashion? Does that need to be part of this document?

>     2. If we decide to go with using the feature test macros, the macros must _not_ be defined by each source file using the functions covered by such macros; instead, these macros should be defined by the libc integration headers under `lib/libc/*/include`. The rationale is that the feature test macros like `_GNU_SOURCE` are not standardised and may vary across different LIBCs, and we do not want to clutter the general source files in the Zephyr codebase with these LIBC-specific feature test macros.

I think all of these feature test macros *are* well defined in their respective standards. `_POSIX_C_SOURCE` is extensively described in the POSIX specs (https://pubs.opengroup.org/onlinepubs/9699919799/). Similarly, `_GNU_SOURCE`  is described in each Linux manual page for the affected functions (see the Glibc `qsort_r()` manual for an example: https://man7.org/linux/man-pages/man3/qsort_r.3.html) along with the general glibc documentation https://ftp.gnu.org/old-gnu/Manuals/glibc-2.2.3/html_node/libc_13.html. Unfortunately, `_BSD_SOURCE` is not described in any BSD documentation I can find, only in the glibc docs. The BSD community doesn't appear to be interested in having a C library provide a way to express different APIs? That's a big part of why I think using them directly in the C source file is a good idea -- that's how those functions are documented to be used.

Sadly, you cannot place them anywhere else -- you cannot change the C library API once the first header file has been processed so the way these feature macros work is incompatible with placing them in a header file.

Hrm. Now I realize a flaw in my plan -- a single C source file can only use a single API definition. If you define both `_POSIX_C_SOURCE` and `_GNU_SOURCE` in a source file, the glibc, picolibc and newlib will all expose the GNU API to the application. This means you cannot have both the POSIX `strerror_r()` and GNU `qsort_r()` functions accessible at the same time.

> If you are referring to the changes in #57619, the `lib/libc/picolibc/include/*.h` providing the Zephyr-side prototype declarations for the non-standard functions is only a temporary workaround until we update the minimum required Zephyr SDK version to a version that contains a more up-to-date Picolibc with `_ZEPHYR_SOURCE` support -- once the minimum required Zephyr SDK version is updated, the Picolibc workarounds can be removed entirely and we can rely on `_ZEPHYR_SOURCE`.

I think it will be a permanent workaround to handle version skew between the picolibc module and the SDK as functions are added in the process defined by Rule A.5.

> Are you saying that, for instance, we define our own `string.h` with all standard and non-standard `string.h` functions so that we do not need to include the LIBC `string.h` at all?

Not quite -- create a zephyr/string.h include file that makes sure all of the extended function declarations are available, perhaps also doing a `#include <string.h>` to pull in the API available from the underlying C library. When using the common C library function, the extended function declarations could be in this file, or perhaps in a file provided by the common C library, possibly protected by a `_*_SOURCE` check so that they weren't available unless the application requested them. When using another C library, this file wouldn't need to do anything extra as the C library string.h would be expected to provide the requested API.

stephanosio replied:
> Not entirely -- we will need to add support in Zephyr to support older SDK versions and/or external versions of said libraries. Just as #57619 is adding picolibc support.

Once we make Picolibc the default libc for Zephyr, I think it would be reasonable to update the minimum required Zephyr SDK version when a change is made to the Rule A.5 (and subsequently the Picolibc `_ZEPHYR_SOURCE`), in which case we would not need hacks like #57619 for Picolibc.

> Sounds like there's a requirement to add tests within Zephyr for all functions added in this fashion? Does that need to be part of this document?

Yes, there needs to be an associated test when you introduce a new feature to Zephyr -- this is more or less a general project-wide requirement than a libc-specific one.

> I think all of these feature test macros _are_ well defined in their respective standards. `_POSIX_C_SOURCE` is extensively described in the POSIX specs (https://pubs.opengroup.org/onlinepubs/9699919799/). Similarly, `_GNU_SOURCE` is described in each Linux manual page for the affected functions (see the Glibc `qsort_r()` manual for an example: https://man7.org/linux/man-pages/man3/qsort_r.3.html) along with the general glibc documentation https://ftp.gnu.org/old-gnu/Manuals/glibc-2.2.3/html_node/libc_13.html. Unfortunately, `_BSD_SOURCE` is not described in any BSD documentation I can find, only in the glibc docs. The BSD community doesn't appear to be interested in having a C library provide a way to express different APIs? That's a big part of why I think using them directly in the C source file is a good idea -- that's how those functions are documented to be used.

Neither the Linux man page nor the glibc documentation is a "standard." Not to mention the `_BSD_SOURCE` is not clearly defined anywhere as you pointed out. The only real "standard" here is the POSIX for `_POSIX_C_SOURCE`.

In that sense, it is completely up to each LIBC to decide whether they want to respect `_GNU_SOURCE` and `_BSD_SOURCE` (e.g. a LIBC may very well decide to use `__USE_QSORT_R` instead of `_GNU_SOURCE`), and we do not want to end up defining both `_GNU_SOURCE` and `__USE_QSORT_R` in every C file that uses these functions.

> Sadly, you cannot place them anywhere else -- you cannot change the C library API once the first header file has been processed so the way these feature macros work is incompatible with placing them in a header file.
> 
> Hrm. Now I realize a flaw in my plan -- a single C source file can only use a single API definition. If you define both `_POSIX_C_SOURCE` and `_GNU_SOURCE` in a source file, the glibc, picolibc and newlib will all expose the GNU API to the application. This means you cannot have both the POSIX `strerror_r()` and GNU `qsort_r()` functions accessible at the same time.

That is indeed a problem, as we have already seen. At that point, unless we modify the LIBC to support `_ZEPHYR_SOURCE` or handle the conflicts otherwise, I believe the best we can do is what is done in #57619.

> I think it will be a permanent workaround to handle version skew between the picolibc module and the SDK as functions are added in the process defined by Rule A.5.

We only need to keep that workaround until we update the minimum required Zephyr SDK version, and as per the pointed I made above, I think it is reasonable to update the minimum required Zephyr SDK version when a change is made to the Rule A.5 and subsequently the Picolibc `_ZEPHYR_SOURCE`.

> Not quite -- create a zephyr/string.h include file that makes sure all of the extended function declarations are available, perhaps also doing a `#include <string.h>` to pull in the API available from the underlying C library. When using the common C library function, the extended function declarations could be in this file, or perhaps in a file provided by the common C library, possibly protected by a `_*_SOURCE` check so that they weren't available unless the application requested them. When using another C library, this file wouldn't need to do anything extra as the C library string.h would be expected to provide the requested API.

I am still 100% clear on what you are suggesting here. How is `(include/)zephyr/string.h` supposed to be included? Are you saying that Zephyr application should include `zephyr/string.h` instead of `string.h`?

Maybe a more discrete example would help.

keith-packard replied:
> Neither the Linux man page nor the glibc documentation is a "standard." Not to mention the `_BSD_SOURCE` is not clearly defined anywhere as you pointed out. The only real "standard" here is the POSIX for `_POSIX_C_SOURCE`.

We could argue semantics about what makes a standard, but given that the goal is for Zephyr to be able to mix-n-match functions from a variety of sources, it's clear it won't matter in the end -- we don't know anything better than #57619 that can be used to describe the Zephyr API for C libraries that don't provide `_ZEPHYR_SOURCE` support.

> We only need to keep that workaround until we update the minimum required Zephyr SDK version, and as per the pointed I made above, I think it is reasonable to update the minimum required Zephyr SDK version when a change is made to the Rule A.5 and subsequently the Picolibc `_ZEPHYR_SOURCE`.

I suspect there will be an occasional need to support multiple picolibc versions for Zephyr, whether to allow use of a range of SDK versions or to allow use of an external toolchain with picolibc support (like armclang).

If we can add validation of the non ISO/IEC C library APIs provided by `_ZEPHYR_SOURCE` using the technique I suggested in #57619, then I think we'll catch API mismatches during testing, which reduces the concerns I had over adding C library specific declarations for functions within the Zephyr code base.

keith-packard commented:
Oof. I just learned about `strnlen_s`, which is defined in the C11 spec, but requires that the implementation support functions in annex K. And I haven't found any C library which even talks about this. Yay for having so many standards to choose from!

stephanosio commented:
> Oof. I just learned about `strnlen_s`, which is defined in the C11 spec, but requires that the implementation support functions in annex K. And I haven't found any C library which even talks about this. Yay for having so many standards to choose from!

In the past, we discussed adopting the optional Annex K alongside the safeclib in the Zephyr codebase, but it was rejected by the Security WG on the grounds that the Annex K is too controversial.

stephanosio commented on doc/contribute/coding_guidelines/index.rst:
```diff
@@ -1297,6 +1297,188 @@ macros. Given that Zephyr uses a fork of the corresponding upstream for each
+Description
+-----------
+
+The use of the C standard library functions and macros in the Zephyr codebase
+shall be limited to the functions from the ISO/IEC 9899:2011 standard, also
+known as C11, unless exempted by this rule.
```
@keith-packard Speaking of the _optional_ features in https://github.com/zephyrproject-rtos/zephyr/pull/57598#issuecomment-1550304383, I should probably revise the Rule A.5 to refer to the _mandatory_  functions and macros.

```suggestion
The use of the C standard library functions and macros in the Zephyr codebase
shall be limited to the mandatory functions from the ISO/IEC 9899:2011 standard, also
known as C11, unless exempted by this rule.
```

keith-packard replied:
Yeah, we definitely don't want to stick annex K in here. We probably do want `<math.h>` though, and that's in annex F (given the initial, one suspects a bit of careful annex ordering here), and probably `<complex.h>`, which is in annex G. Maybe just exclude annex K?

stephanosio replied:
Done. https://github.com/zephyrproject-rtos/zephyr/pull/57598#issuecomment-1553424876

stephanosio commented:
The general direction of making Picolibc the default libc for Zephyr and introducing the new coding guidelines rules to regulate the C standard library function usage have been approved by the TSC on 2023-05-17.

The exact phrasing of the coding guidelines rules still need to be finalised before this PR can be merged.

stephanosio commented:
**UPDATES**

* Removed `qsort_r()`, `reallocarray()`, `strerror_r()` and `strtok_r()`, which are not currently used anywhere inside the "Zephyr kernel" scope, from the Rule A.4.
* Removed `qsort_r()`, `reallocarray()` and `strerror_r()`, which are not currently used anywhere in the Zephyr codebase, from the Rule A.5. Note that `strtok_r()` was left in because it is currently used outside the "Zephyr kernel" scope.
* Added a comment about introducing new functions in the Rule A.4 and Rule A.5.
* Added a comment to clarify that the functions listed in the Rule A.4 are not _unconditionally_ allowed to be used in the "Zephyr kernel" scope.
* Clarified the scope of ISO/IEC 9899:2011 functions in the Rule A.5 to exclude the Annex K.

stephanosio commented:
@keith-packard @cfriedt All of the concerns raised earlier should be addressed now. Please take another look.

cfriedt commented on doc/contribute/coding_guidelines/index.rst:
```diff
@@ -1297,6 +1297,191 @@ macros. Given that Zephyr uses a fork of the corresponding upstream for each
+
+   abort(),ISO/IEC 9899:2011
+   abs(),ISO/IEC 9899:2011
+   aligned_alloc(),ISO/IEC 9899:2011
```
@stephanosio - iirc the items in this list are permitted (although in some cases additional justification should be made).

Just curious though - does it make sense to use std c allocation and free functions inside the kernel, or would it be better to insist that k_ variants be used for the kernel?

My thoughts are that the kernel and most native zephyr components should use k_ variants for heap allocations and frees.

It obviously makes sense to provide the std libc interface for allocation as a compatibility layer with libraries, etc.

Just thought I would mention - it's probably still covered in the sense that any use is technically justified.

stephanosio replied:
> Just curious though - does it make sense to use std c allocation and free functions inside the kernel, or would it be better to insist that k_ variants be used for the kernel?

It is always recommended to use the `k_malloc` instead of the libc `malloc` for Zephyr native applications, including the kernel itself -- that is actually documented here:

https://github.com/zephyrproject-rtos/zephyr/blob/6697c5aa0bb296ad008ef8f728df254954764962/doc/develop/languages/c/index.rst?plain=1#L112-L120

Also I have already addressed the concern raised in https://github.com/zephyrproject-rtos/zephyr/pull/57598#issuecomment-1550089122 and added the following:

```
Note that the use of the functions listed above are subject to secure and safe
coding practices and it should not be assumed that their use in the Zephyr
kernel is unconditionally permitted by being listed in this rule.
```

keith-packard replied:
I thought the `k_malloc` family could only be used in kernel mode, not userspace, and that was why there was a separation between it and `malloc`. Otherwise, the common C library would have just used `k_malloc` to implement `malloc', right?

stephanosio replied:
> I thought the `k_malloc` family could only be used in kernel mode, not userspace

Correct. The exact recommendation is to use the "memory management API," which `k_malloc` is part of.

> that was why there was a separation between it and `malloc`.

No, the `malloc` is there for supporting portable applications using the libc API. If targeting userspace, Zephyr native applications should define their own memory domains/partitions/heap, use the applicable userspace-safe APIs (e.g. `sys_heap`).

keith-packard replied:
Yes, Zephyr native code would use native memory APIs. My point was that as `malloc()` needs to support non-native code in both kernel and userspace allocations and so cannot simply use `k_malloc()`.

stephanosio replied:
> Yes, Zephyr native code would use native memory APIs. My point was that as `malloc()` needs to support non-native code in both kernel and userspace allocations and so cannot simply use `k_malloc()`.

Yes.

keith-packard commented on doc/contribute/coding_guidelines/index.rst:
```diff
@@ -1297,6 +1297,191 @@ macros. Given that Zephyr uses a fork of the corresponding upstream for each
+   :header: Function,Source
+   :widths: auto
+
+   abort(),ISO/IEC 9899:2011
```
You've defined C11 as an alias for ISO/IEC 9899:2011, might be nice to use the shorter name in this table?

stephanosio replied:
Given that there is plenty of horizontal space left when this table gets rendered, I would prefer to keep the full name.

It would also be nicer to use the full name here from the consistency perspective in case we ever need to reference other standards in the future.

keith-packard reviewed (approved):
This looks good to me now.

stephanosio commented:
Merging since this has been approved by both the TSC and the Coding Guidelines maintainer.
